lib/api/core/plugins/pluginsManager.js

```diff
@@ -238,15 +237,406 @@ class PluginsManager {
+        throw new PluginImplementationError(error);
+      }
+    });
+  }
+
+  /**
+   * Register a listener function on an event
+   *
+   * @param {object} plugin
+   * @param {string} event
+   * @param {string} fn - function to attach
+   */
+  registerHook(plugin, event, fn) {
+    debug('[%s] register hook on event "%s"', plugin.name, event);
+
+    this.kuzzle.on(event, message => {
+      try {
+        plugin.object[fn](message, event);
+      }
+      catch (error) {
+        throw new PluginImplementationError(error);
+      }
+    });
+  }
+
+  /**
+   * Registers an authentication strategy.
+   *
+   * @param {string} pluginName - plugin name
+   * @param {string} strategyName - strategy name
+   * @param {object} strategy - strategy properties
+   * @throws {PluginImplementationError} If the strategy is invalid or if registration fails
+   */
+  registerStrategy(pluginName, strategyName, strategy) {
+    this.validateStrategy(pluginName, strategyName, strategy);
+
+    const 
+      plugin = this.plugins[pluginName],
+      methods = {};
+
+    // wrapping plugin methods to force their context and to 
+    // cast uncaught errors into PluginImplementationError errors
+    Object.keys(strategy.methods).filter(name => name !== 'verify').forEach(methodName => {
+      methods[methodName] = (...args) => {
+        return new Bluebird((resolve, reject) => {
+          try {
+            resolve(plugin.object[strategy.methods[methodName]].bind(plugin.object)(...args));
+          }
+          catch (error) {
+            reject(error instanceof KuzzleError ? error : new PluginImplementationError(error));
+          }
+        });
+      };
+    });
+
+    const 
+      opts = Object.assign(strategy.config.strategyOptions || {}, {passReqToCallback: true}),
+      verifyAdapter = (...args) => {
+        const callback = args[args.length - 1];
+
+        return plugin.object[strategy.methods.verify](...args.slice(0, -1))
+          .then(result => {
+            if (result && typeof result === 'object') {
+              if (result.kuid && typeof result.kuid === 'string') {
+                this.kuzzle.repositories.user.load(result.kuid)
+                  .then(user => {
+                    if (user === null) {
+                      callback(new PluginImplementationError(`The strategy "${strategyName}" returned an unknown Kuzzle user identifier.`));
+                    }
+                    else {
+                      callback(null, user);
+                    }
+                  })
+                  .catch(error => callback(error));
+              }
+              else if (result.message && typeof result.message === 'string') {
+                callback(null, false, result);
+              }
+              else {
+                callback(null, false, `Was not able to log in using the strategy "${strategyName}"`);
+              }
+            }
+            else {
+              callback(new UnauthorizedError('Login failed.'));
```

Comment:
Yes and no.
You're right about plugins returning `false` as a result, as it should be handled as a simple login rejection, and not an exception.

But we also should handle cases where plugins return erroneous values (for instance, an array, a number, etc).

So I added a new test case: if `result` is explicitly set to `false` then we invoke the callback with that boolean value. In all other cases, we invoke the callback with a `PluginImplementationError` exception (instead of an `UnauthorizedError` one).